add rand(::String)

rfourquet opened the pull request:
```
julia> rand("asd", 3)
3-element Array{Char,1}:
 'd'
 'a'
 'a'
```

This should work more generally for any iterable (with `length`) in principle, but I'm hesitant to introduce a method of `rand` with unconstrained parameters: `rand([rng], iter)`, which would mean that this would be the fallback method anytime `rand` is called on a type for which it's not defined (so maybe this should wait till some traits for iterators are available).

rfourquet committed:
add rand(::String)

StefanKarpinski commented on test/random.jl:
```diff
@@ -313,7 +313,9 @@ for rng in ([], [MersenneTwister(0)], [RandomDevice()])
-                   (rand(Int, 100), Int)]
+                   (rand(Int, 100), Int),
+                   ("qwèrtï", Char),
+                   (Base.Test.GenericString("qwèrtï"), Char)]
```
I would throw in a test for empty strings since that seems like a corner case that could easily break. (I checked this out specifically to test that – it works.)

rfourquet committed:
test for empty strings (StefanKarpinski) and fix bad usage a GLOBAL_RNG

ararslan commented on base/random.jl:
```diff
@@ -683,6 +683,44 @@ end
+
+rand_iter(rng, s, ::Base.IteratorSize) = throw(ArgumentError("iterator must have a known length"))
+
+function rand_iter(rng::AbstractRNG, s, ::Union{Base.HasShape,Base.HasLength})::eltype(s)
```
Out of curiosity, why is the return type annotation necessary here? `c` is the value that will be returned and isn't that already guaranteed to have type `eltype(s)` (or at least `<:eltype(s)` if `eltype(s)` is abstract) given that it's an element of `s`?

tkelman replied:
inference probably can't determine that `i == pos` is guaranteed to be true for at least one value of the enumerate loop

rfourquet replied:
Yes Tony is spot on, and also initially I thought to propose this PR working generally on iterators, and the return type annotation could serve as an additional way to check whether the argument was complying to the iterator interface.

ararslan replied:
Okay, nice. Thanks for the explanation, both of you.

stevengj commented on base/random.jl:
```diff
@@ -683,6 +683,44 @@ end
+    for (i, c) in enumerate(s)
+        i == pos && return c
+    end
+end
```
I'm a little skeptical of the `rand_iter` function and `AbstractString` support.    People using `rand` with other string types will probably not expect this function to be O(n); it seems preferable to throw a MethodError rather than to use this implementation.

rfourquet replied:
Yes I agree, I had the same hesitation and had planned to update the doc today to warn about the complexity. Then I think that with a proper warning those convenience methods (similarly for `AbstractSet` and `Associative` at #22228) are not harmful and can be handy.

StefanKarpinski replied:
You could also do rejection sampling: randomly choose and index from `start(s):endof(s)` and retry if it's not a valid start of a character – i.e. the same thing you do for `String`.

rfourquet replied:
Oh great! for some reason, I skimmed too fast on the generic code of `isvalid` and thought it was O(n).

rfourquet replied:
Actually, I was confused not about `isvalid`, but about the fact that I thought I needed to compute the `length` somehow, which is O(n) in the generic method. Thanks again @StefanKarpinski, I updated the code.

rfourquet committed:
warn about linear complexity for AbstractString (stevengj)

ararslan commented on base/random.jl:
```diff
@@ -257,13 +257,19 @@ globalRNG() = GLOBAL_RNG
 
 `S` defaults to [`Float64`](@ref).
 
+Note that the complexity of `rand(rng, s::AbstractString)` is linear
```
Not really necessary, but this could be a Documenter `note` block, like
```
!!! note
    The complexity of `rand(rng, s::AbstractString)` is linear in the
    length of `s` if `s` is not of type `String`. If called more than
    a few times, use `rand(rng, collect(s))` instead.
```

stevengj replied:
and maybe `rand(rng, collect(s)) or rand(rng, String(s))`, just to emphasize that `String` is fine.

stevengj replied:
"If called more than a few times" is a bit awkward because of the ambiguous subject of "called".   Maybe "For generating many random characters, use ..."

rfourquet replied:
I understand, but your new formulation seems to include also the case were I call `rand(rng, s, dims)` which already does a collect... WDYT?

stevengj replied:
Maybe "For multiple calls, consider using ..."

rfourquet committed:
use O(1) algo for generic AbstractString (StefanKarpinski)

rfourquet commented on base/random.jl:
```diff
@@ -683,6 +685,39 @@ end
+            return s[rand(rng, g)]
+        end
+    end
+end
```
@StefanKarpinski would you mind having a look at this updated generic implementation, to confirm it complies with `AbstractString` API?

StefanKarpinski replied:
I think it would be better to do `i = rand(rng, g); isvalid(s, i) && return i` since various string types are expected to implement that without the try/catch. The try/catch only exists as a last resort fall-back, and I believe it is more targeted and only catches `UnicodeError` exceptions (if it's not already, it should be).

rfourquet replied:
Ok thanks (I believe the idea being that try/catch is expected to occur, and slow down the computation when the `isvalid` call would generally be faster). I will update. Note that the generic `isvalid` catches all exceptions (not only `UnicodeError`). I'm not familiar with strings, but I can open a corresponding simple PR if it helps.

rfourquet committed:
use isvalid instead of try/catch in generic implementation (StefanKarpinski)

rfourquet commented:
I rebased (wile changing only the docstring, including "a string (considered as a collection of characters)"), and removed the try/catch block in favor of `isvalid`.

stevengj commented on base/random.jl:
```diff
@@ -256,8 +256,9 @@ globalRNG() = GLOBAL_RNG
-* an `Associative` or `AbstractSet` object, or
+* an indexable collection (for example `1:n` or `['x','y','z']`),
+* an `Associative` or `AbstractSet` object,
+* a string (considered as a collection of characters),
```
need "or" on the end of this line

rfourquet committed:
add a missing "or" in doc (stevengj)

rfourquet commented:
Does it need a NEWS.md entry?

rfourquet committed:
add NEWS.md entry

rfourquet commented:
Thanks all for your help!
